### ISSUE TITLE
ODATA: How to get list of users using '@odata.nextLink'

### ISSUE THREAD

**Issue by madanbisht** [davidmoten/odata-client#15]
Hi David,
       We appreciate that odata client library is a powerful tool to get data from exchange server. Currently in our use case the challenge is to get the list of users, given the value of '@odata.nextLink'.

As an example below MS Graph Java SDK code can be used to get list of users based on '@odata.nextLink'.
```
IUserCollectionRequestBuilder userCollReqBuilder = new UserCollectionRequestBuilder(@odata.nextLink, graphClient, null);
if (userCollReqBuilder != null) {
    IUserCollectionPage response = userCollReqBuilder.buildRequest().get();
    List<User> users = response.getCurrentPage();
}
```

 It will be great if you can help us to get user list using '@odata.nextLink'.

Thanks
Madan

**Comment by davidmoten** [comment]
Why do you care about low-level stuff like `@odata.nextLink`? Just use the collections support in the API. See [Collections and paging](https://github.com/davidmoten/odata-client#collections-and-paging) in the README.md. Can you explain your use case further?

**Comment by madanbisht** [comment]
Thanks for your response and below is the use-case:

There is a service which receives request from a client and we are trying to implement paging, so that the service will send only the list of users requested by the client. As an example:

**Step 1)** Client send request to get the list of N users and service can get user-list as below:
```
     OdataClientFactory
                    .request()
                    .groups().top(N)
                    .get();
```
> After getting response client will parse N users and extract @data.nextLink from the response.

**Step 2)** Client sends request to get next N users providing the @data.nextLink and service can use @data.nextLink to get next N users.

> After getting response client will parse the user-list and extract @data.nextLink from the response.

Step 3) Repetition of step 2.

**Comment by davidmoten** [comment]
So you're sort of proxying the Graph API for the client? @odata.nextLink just uses a skip field. You could add a counter field to the returned JSON and then get the user to pass the last counter value to the service instead of `@odata.nextLink` and use 

```java
List<User> page = client.client.users().skip(counter).get().currentPage();
```
when they call the service. That way you are not coupled to odata metadata like `@odata.nextLink`.

I could probably provide access to the '@odata.nextLink` field, I just wanted to explore alternatives first.

**Comment by madanbisht** [comment]
We already tried **skip** option(as given below) but as soon as we use **skip** it throws an exception(java.lang.NullPointerException). It seems **skip** is not supported.
```
List<User> users = OdataClientFactory
                    .request()
                    .users().skip(1)
                    .get()
                    .currentPage();
```
Ideally our use case can be implemented with below code, here **top** works fine but as soon as we use **skip** it throws exception.
```
List<User> users = OdataClientFactory
                    .request()
                    .users().top(200).skip(100)
                    .get()
                    .currentPage();
```

**Comment by davidmoten** [comment]
Can you supply full stack trace?

**Comment by davidmoten** [comment]
By the way if `top` is your attempt to control page size I don't think it works.

**Comment by davidmoten** [comment]
I just tried `client.users().top(200).skip(100).get().currentPage()` and like you I got an NPE. I'll fix that. However the base cause is 
```
10:44:32.468 [main] DEBUG com.github.davidmoten.odata.client.internal.ApacheHttpClientHttpService - response text=
{
  "error": {
    "code": "Request_BadRequest",
    "message": "'$skip' is not supported by the service.",
    "innerError": {
      "date": "2020-06-13T00:44:32",
      "request-id": "63ed6165-729f-4753-88b9-e67399a312f5"
    }
  }
}
```

**Comment by davidmoten** [comment]
In fact, digging a bit deeper I don't even have privileges:

```java
client.users().stream().forEach(user -> System.out.println(user.getPrincipalName()));
```
gives an NPE (something else to fix!) and log says:
```
{
  "error": {
    "code": "Authorization_RequestDenied",
    "message": "Insufficient privileges to complete the operation.",
    "innerError": {
      "date": "2020-06-13T00:48:55",
      "request-id": "31c91f39-f97b-4f8b-8f6f-c4c00c1a8fe9"
    }
  }
}
```

**Comment by davidmoten** [comment]
Reading a bit further I think the collection returned by `users().get()` must use $skipToken for pagination instead of $skip. Given this I need to rethink the advice about using `skip`.

**Comment by davidmoten** [comment]
Please review PR #17.

**Comment by davidmoten** [comment]
I've addressed this issue now with #17. 

You can specify the link for a collection like this:

```java
List<User> users = client.users().urlOverride(url).get().currentPage();
```
I'll release shortly, let me know if any problems.

**Comment by davidmoten** [comment]
0.1.16 released to Maven Central with #16, #17 that resolve this issue. You may wish to watch this repository to hear about future releases (which are pretty frequent lately).

**Comment by madanbisht** [comment]
Hi Dave,
We have verified and its working perfectly. 

We appreciate your consideration and time on developing this to us so quickly.

Thanks
Madan
